### ISSUE TITLE
Questions about optimizer in 6th tutorial (Transformer)

### ISSUE THREAD

**Issue by gorka96** [bentrevett/pytorch-seq2seq#61]
Hi Ben,

I was updating my transformer model code with your last version of the transformer of the 6th tutorial (BTW, I really appreciate that you added attention display to the tutorial), and I got some questions about the implementation.

1: What was the reason to change the optimizer from Noam to Adam? (if there was any)

2: Should the learning rate of the optimizers change when we change the size of the model (hid dim, n layers, n heads, ...) ?

Besides, I was gonna ask if it wasn't missing a `model.eval()` before calling to `evaluate()` function, but I see that you add it inside the function itself, which makes more sense.

Moveover, I had added early_stopping for training and minimum length for the inference (to avoid empty translations, as i have seen in some implementations as in [OpenNMT](https://github.com/OpenNMT/OpenNMT-py/pull/496)) in my code, but its not fancy code. They are easy to implement and useful features, I will try to add them on a pull request when I take the time to rewrite the code.  

Thanks for sharing this well implemented and even better commented models with us.

Gorka

**Comment by bentrevett** [comment]
I have been working on this last tutorial on and off for some time. There's a lot of stuff to explain in it, with all the different layers and how they work/interact so it's taken some time to organise the cells in a way in which I like, but that involved cutting two things out - the sin/cos positional embeddings and the Noam optimizer. 

I decided to cut these two out as BERT, which most transformer models are now based off, does not use either of them. After removing these two it took a long time to get a hyperparameter setting that trained well - however I managed to find one eventually, which also had less parameters and trained considerably faster. 

So, for (1), mainly simplicity but also because modern transformer models, aka BERT, do not use it either. For (2), yes the learning rate should be tuned, but try with the value from the tutorials first as I think it should be decent.
